### DISCUSSION TITLE
VS2017 + conan usage

### DISCUSSION THREAD

**Discussion by malachib** [skypjack/entt#606] [Build system]
Hey, firstly thank you for this amazing library!

Although this is primarily a conan issue, I'm hoping you have some expertise to help me out because the conan discussions go on for miles on the topic.

I am unable to do a `conan install` when using VS2017 due to:

```
ERROR: entt/3.5.2: Invalid configuration: entt requires a compiler that supports at least C++17
```

This is primarily because compiler.version=15 was the first release of VS2017 which had no C++17, but now it does.  Conan has no notion of compiler.version=15.9, and conan's `cppstd=17` isn't noticed because the compiler test fails first.

Do you have any suggestions for using VS2017 + conan + C++17?  It seems a shame to go down the C++14 route (though I am doubly thankful, that you've provided that option)

**Comment by skypjack** [comment]
I've just enabled EnTT discussions [here](https://github.com/skypjack/entt/discussions) (finally, thanks GH!) and this questions is probably better suited for that.
Though, let's continue here anyway. Unfortunately, I'm not an expert of Conan, so I don't know how we can solve it but I'm interested in knowing this.
Maybe @Lawrencemm knows it? He uses Conan afaik and is by far more skilled than me.

**Comment by Lawrencemm** [comment]
Since this package doesn't need to be built, you can do a separate install: `conan install entt/3.5.2 -s compiler.version=6` and it should work as a workaround until you find a better solution. (don't quote me on that command, I'm pretty sure you can install a single package but I can't remember the exact invocation)

**Comment by malachib** [comment]
The trouble we run in to is conan itself performs a compiler check during cmake scan.  The conan VS/MSVC version managing is complicated to say the least.  At any rate, at this point cmake halts because it:

"Detected a mismatch for the compiler version between your profile settings and CMake"

Which is actually technically true, since `compiler.version=16` is VS2019

A workaround seems possible by using `CONAN_DISABLE_CHECK_COMPILER` but, kinda ew.  Will try that and share my findings

_UPDATE_

Am met with error `error C1007: unrecognized flag "-pdbrpc" in "p2"` which is likely VS2017 attempting to use VS2019 link flags.  Described more in detail here https://stackoverflow.com/questions/54664134/cannot-compile-unknown-flag-pdbrpc-in-p2
Also a ton of C4307 warnings about integral constant overflow.  That's all kind of a rabbit hole.  

Rather than going that way, I wonder - would be possible for entt packaging to be upgraded to permit `compiler.version=15` if `cppstd=17` is in effect?

**Comment by Lawrencemm** [comment]
Umm are you using the `cmake` generator for conan? If so I recommend switching to the `cmake_find_package` generator, it doesn't have these weird interactions and decouples the package manager from the build system.

Just remember to set the CMAKE_MODULE_PATH to the build directory when doing so, so that cmake finds the scripts conan generates.

**Comment by skypjack** [comment]
@malachib I'm more than willing to do what is needed in `EnTT` but you've to guide me because I've no clue of what is wrong and how to make it work. 🙂

**Comment by malachib** [comment]
Yeah, I am using the `cmake` generator.  Giving `cmake_find_package` a try, it looks like a big improvement.

_UPDATE_

I've reworked things to use the suggested `cmake_find_package` generator.  It's a universal improvement, but all the aforementioned compile/link issues persist.

It's a little mysterious to me what piece of conan is actually generating that `requires a compiler` message in the first place.  This link: https://stackoverflow.com/questions/64694269/getting-uwebsockets-for-visual-2017-via-conan is seemingly exactly our issue, but the big difference is their conanfile.py actually includes that error message - where we quite reasonably do not.  I figure if I can find the chunk of code generating it, maybe can work backwards to accommodate its wishes.  Closest I can find is:

https://github.com/conan-io/conan-center-index/blob/master/recipes/cpp-taskflow/all/conanfile.py but frankly that seems like a random shot in the dark.

I'm confused 😕

**Comment by skypjack** [comment]
What about contacting someone from Conan for this? Maybe they can provide a solution that we don't see.

**Comment by Lawrencemm** [comment]
@malachib I'd be happy to help further, maybe move this to the discussions area or the discord so we can keep the EnTT issues clean 😅

**Comment by jgsogo** [comment]
Hi!

The error comes from this condition in the `conanfile.py` (https://github.com/conan-io/conan-center-index/blob/master/recipes/entt/3.x.x/conanfile.py#L42)... assuming you are using the reference from ConanCenter. The recipe is wrong, you should be able to provide a profile like:

```
[settings]
...
compiler.version=15
compiler.cppstd=17
```

and it should work, Conan would inject the flag to use C++17 and the library will compile... but given the checks in the recipe, that configuration will fail too.

Let me submit a PR to ConanCenter to fix the issue.

**Comment by malachib** [comment]
Creating #612 as suggested 😄

**Comment by skypjack** [comment]
We can directly convert the issue in a discussion. I think it's better, so as to maintain the history.
Don't get offense, but I'm deleting #612 and moving on this one. 🙂 

Here is the last comment by @malachib from the discussion he created:

> I didn't anticipate that in conan's own github repo there would be some entt configurations! Fascinating
> @skypjack @Lawrencemm I appreciate how much you two dug into this, knowing full well it was a conan-sided thing all along

**Comment by malachib** [comment]
Relinking https://github.com/conan-io/conan-center-index/pull/3840 for fun and profit
Really digging `cmake_find_packages` !

**Reply by Lawrencemm** [reply]
Ah that's good to hear @malachib. @jgsogo thank you very much for helping us out

**Comment by skypjack** [comment]
So, the conan-sided issue is solved as far as I can tell and `EnTT` works correctly also when using conan with VS2017.
Can you confirm this @malachib (because I didn't try it actually 🙂)? Thanks!

**Comment by malachib** [comment]
Been so slammed!!
Yes it works
Thank you!
